Refactor qconv_prepack and qconv_unpack to support conv3d (#146)

xiaomengy opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/FBGEMM/pull/146

Refactor qconv_prepack and qconv_unpack to support conv3d

Test Plan: buck test mode/dev-nosan caffe2/test:quantized -- "Conv"

Reviewed By: dskhudia

Differential Revision: D18023651

BIT-silence commented:
link to https://github.com/pytorch/pytorch/issues/28480

BIT-silence commented:
link to https://github.com/pytorch/FBGEMM/pull/146

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -263,12 +294,86 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+
+static auto registry =
+    c10::RegisterOperators()
+        .op("quantized::conv_prepack",
```
Hmm, I don't really like the naming inconsistency here (`conv_prepack` vs `conv3d_prepack`, i'd prefer `conv2d_prepack`). It would be BC breaking but quantization is experimental so it's probably okay

BIT-silence replied:
The problem is conv_prepack op is already there and if we change the api name, there may be some backward compatibility issues.

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp:
```diff
@@ -103,11 +119,15 @@ class QConvUnpackWeightsInt8 final : public c10::OperatorKernel {
-        TensorTypeId::CPUTensorId));
+static auto registry =
+    c10::RegisterOperators()
+        .op("quantized::conv_unpack(Tensor packed_weights)"
```
Same comment on naming inconsistency as above

BIT-silence replied:
Same as the comments above, because of backward compatibility tests, we cannot change the existing op names and args

jamesr66a replied:
What I’m saying above is we probably *should* break backwards compatibility, otherwise we’re in a state that doesn’t make sense. We can defer this to another PR though for now

jamesr66a commented on test/test_quantized.py:
```diff
@@ -1276,6 +1276,48 @@ def test_qlinear_unpack(self, W, use_channelwise, qengine):
 
 class TestQuantizedConv(unittest.TestCase):
+    def _test_qconv_unpack_impl(
+        self, qconv_prepack, qconv_unpack, inputs, strides, pads, channelwise
```
Naming the function arguments `qconv_prepack` and `qconv_unpack` can be confusing for readers. Can you name them something like `unpack_fn` and `pack_fn` to distinguish them from the ops?

BIT-silence replied:
Done.

jamesr66a commented on c10/core/MemoryFormat.h:
```diff
@@ -42,10 +43,15 @@ inline std::ostream& operator<<(
+  } else {
+    strides[4] = sizes[1];
+    strides[3] = strides[4] * sizes[4];
+  }
```
cc @VitalyFedyunin does this look OK?

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp:
```diff
@@ -93,6 +105,10 @@ class QConvUnpackWeightsInt8 final : public c10::OperatorKernel {
+      TORCH_CHECK(
+          kSpatialDim == 2,
+          "quantized::conv_unpack (qnnpack): QNNPACK only supports Conv2d "
+          "now.");
```
:'( any plan to support it in the future?

BIT-silence replied:
I guess there is? Actually I am not quite sure about that part.

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -263,12 +294,86 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+    const torch::List<int64_t>& stride,
+    const torch::List<int64_t>& padding,
+    const torch::List<int64_t>& dilation) {
+  return fbgemm::conv_param_t<2>(
```
for this `conv_param_t` invocation and the one below, can we getp aram name comments (e.g. like https://github.com/pytorch/pytorch/blob/d83389d327157b04ea74278c1026e6e370ba4001/aten/src/ATen/native/quantized/cpu/qlinear.cpp#L170)

BIT-silence replied:
Done.

jamesr66a reviewed (commented):
Generally looks OK. I'd like to defer to Vitaly to verify the memory layout changes

VitalyFedyunin commented:
CC @VitalyFedyunin to review memory format usage.

VitalyFedyunin commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -31,90 +38,103 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
-        reinterpret_cast<int8_t*>(weight_contig.data_ptr<c10::qint8>());
-
+    // FBGEMM expects weights to be in channels last
+    const Tensor weight_tensor = weight.contiguous(MemoryFormat::ChannelsLast);
```
Will fail for 5d tensor

BIT-silence replied:
Sure, I have make a fallback change here.

VitalyFedyunin commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -0,0 +1,195 @@
+      {dilations[0], dilations[1], dilations[2]}); // dilations
+}
+
+Tensor MakeStridedQTensorCPU(
```
This is nice candidate to be public function similar to `empty_strided`

VitalyFedyunin commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -0,0 +1,195 @@
+        src.quantizer());
+    AT_DISPATCH_QINT_TYPES(
+        src.scalar_type(), "ConvertToChannelsLast3dTensor", [&]() {
+          const auto& src_ref = src.is_contiguous() ? src : src.contiguous();
```
No need to check is_contiguous, src.contiguous() will provide same functionality internally.

BIT-silence replied:
Done.

VitalyFedyunin reviewed (commented):
LGTG for me in terms of memory formats. Took a couple of important notes for me on how to implement 3d channels last.

xiaomengy committed:
Refactor qconv_prepack and qconv_unpack to support conv3d (#28481)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/28481

Refactor qconv_prepack and qconv_unpack to support conv3d

Test Plan: buck test mode/dev-nosan caffe2/test:quantized -- "Conv"

Reviewed By: dskhudia

Differential Revision: D18023651

fbshipit-source-id: 4c6be0e3ebf9d995e442c5be63d4316944fa9440

jianyuh reviewed (approved):
LGTM! @dskhudia has made the comments and approved in the internal tools. Please make sure all OSS tests pass.
